server/commands/list_objects.go

```diff
@@ -0,0 +1,372 @@
+
+				resp, err := query.Execute(timeoutCtx, &openfgapb.CheckRequest{
+					StoreId:              req.StoreId,
+					AuthorizationModelId: req.AuthorizationModelId,
+					TupleKey: &openfgapb.TupleKey{
+						Object:   t.Key.Object,
+						Relation: req.GetRelation(),
+						User:     req.GetUser(),
+					},
+				})
+				if err != nil {
+					return err
+				}
+
+				if resp.Allowed && !uniqueObjectIdsSet.Has(objectID) && atomic.LoadUint32(&objectsFound) < q.listObjectsMaxResults {
+					err = srv.Send(&openfgapb.StreamedListObjectsResponse{
+						ObjectId: objectID,
+					})
+					if err != nil {
+						return err
+					}
+
+					atomic.AddUint32(&objectsFound, 1)
+				}
+
+				return nil
+			}
+
+			g.Go(fn)
+		}
+
+		// iterate over all tuples in the contextual tuples input
+		if req.ContextualTuples != nil {
+			for _, t := range req.ContextualTuples.TupleKeys {
+				if atomic.LoadUint32(&objectsFound) >= q.listObjectsMaxResults {
+					break
+				}
+
+				t := t
+				checkFunction := func() error {
+					_, objectID := tuple.SplitObject(t.Object)
+					if uniqueObjectIdsSet.Has(objectID) {
+						return nil
+					}
+					query := NewCheckQuery(q.datastore, q.tracer, q.meter, q.logger, q.resolveNodeLimit)
+
+					resp, err := query.Execute(timeoutCtx, &openfgapb.CheckRequest{
+						StoreId:              req.StoreId,
+						AuthorizationModelId: req.AuthorizationModelId,
+						TupleKey: &openfgapb.TupleKey{
+							Object:   t.Object,
+							Relation: req.GetRelation(),
+							User:     req.GetUser(),
+						},
+						ContextualTuples: req.ContextualTuples,
+					})
+					if err != nil {
+						return err
+					}
+
+					if resp.Allowed && !uniqueObjectIdsSet.Has(objectID) && atomic.LoadUint32(&objectsFound) < q.listObjectsMaxResults {
+						err = srv.Send(&openfgapb.StreamedListObjectsResponse{
+							ObjectId: objectID,
+						})
+						if err != nil {
+							return err
+						}
+
+						atomic.AddUint32(&objectsFound, 1)
+					}
+
+					return nil
+				}
+
+				g.Go(checkFunction)
+			}
+		}
+
+		err := g.Wait()
+		if err != nil {
+			// we don't want to abort everything if one of the checks failed.
+			q.logger.Warn("ListObjects errored: ", logger.Error(err))
+		}
+
+		resolvedChan <- struct{}{}
+		close(errChan)
+	}()
+
+	select {
+	case <-timeoutCtx.Done():
+	case <-resolvedChan:
+	case genericError := <-errChan:
+		return genericError
+	}
+
+	return nil
+}
+
+func (q *ListObjectsQuery) getTupleIterator(ctx context.Context, storeID, targetObjectType string) (storage.TupleIterator, error) {
+	iter, err := q.datastore.ReadRelationshipTuples(ctx, storage.ReadRelationshipTuplesFilter{
+		StoreID:            storeID,
+		OptionalObjectType: targetObjectType,
+	})
+	if err != nil {
+		return nil, serverErrors.NewInternalError("", err)
+	}
+	return iter, nil
+}
+
+func (q *ListObjectsQuery) validateInput(ctx context.Context, storeID string, targetObjectType string, authModelID string, relation string) error {
```

Comment:
Why is this not using https://github.com/openfga/openfga/blob/main/storage/storage.go#L89?

Reply:
Good point, will refactor.